File: examples/incident/client/patches.go
package client

import (
	deep "github.com/brunoga/deep/v6"

	"github.com/brunoga/deep/examples/incident/model"
)

// The patch vocabulary: every write commandpost performs, as a named patch
// constructor. Conditions ride inside the patches, so the concurrency story
// needs no locks anywhere — the server evaluates each condition against the
// state the patch actually meets.

func statusPath() deep.Path[model.Incident, model.Status] {
	return deep.PathString[model.Incident, model.Status]("/status")
}

func severityPath() deep.Path[model.Incident, model.Severity] {
	return deep.PathString[model.Incident, model.Severity]("/severity")
}

func taskField(taskID, field string) string {
	return "/tasks/" + deep.EscapePathKey(taskID) + "/" + field
}

// SetTitle renames the incident.
func SetTitle(title string) deep.Patch[model.Incident] {
	return deep.NewPatch[model.Incident]().With(
		deep.Set(deep.PathString[model.Incident, string]("/title"), title),
	).Build()
}

// SetStatus moves the incident to a status, unconditionally; the server still
// validates the outcome (closed with open tasks is refused there).
func SetStatus(s model.Status) deep.Patch[model.Incident] {
	return deep.NewPatch[model.Incident]().With(deep.Set(statusPath(), s)).Build()
}

// Escalate raises severity to sev — but only If the incident is currently
// less severe. Applied twice, or after someone escalated further, it skips:
// an escalation can never de-escalate, no matter how late it arrives.
func Escalate(sev model.Severity) deep.Patch[model.Incident] {
	return deep.NewPatch[model.Incident]().With(
		deep.Set(severityPath(), sev).If(deep.Gt(severityPath(), sev)),
	).Build()
}

// SetSeverity sets severity outright — the unconditional counterpart, for
// deliberate downgrades.
func SetSeverity(sev model.Severity) deep.Patch[model.Incident] {
	return deep.NewPatch[model.Incident]().With(deep.Set(severityPath(), sev)).Build()
}

// ClaimTask takes a task — If nobody holds it. Two responders racing to
// claim resolve atomically on the server: one applies, one skips.
func ClaimTask(taskID, who string) deep.Patch[model.Incident] {
	owner := deep.PathString[model.Incident, string](taskField(taskID, "owner"))
	return deep.NewPatch[model.Incident]().With(
		deep.Set(owner, who).If(deep.Eq(owner, "")),
	).Build()
}

// CompleteTask marks a task done.
func CompleteTask(taskID string) deep.Patch[model.Incident] {
	return deep.NewPatch[model.Incident]().With(
		deep.Set(deep.PathString[model.Incident, bool](taskField(taskID, "done")), true),
	).Build()
}

// AddTask appends a task to the checklist, addressed by its key — the keyed
// list makes /tasks/<id> a stable name whatever position it lands in.
func AddTask(t model.Task) deep.Patch[model.Incident] {
	return deep.Patch[model.Incident]{Operations: []deep.Operation{{
		Kind: deep.OpAdd,
		Path: "/tasks/" + deep.EscapePathKey(t.ID),
		New:  t,
	}}}
}

// RemoveTask deletes a task by key.
func RemoveTask(taskID string) deep.Patch[model.Incident] {
	return deep.Patch[model.Incident]{Operations: []deep.Operation{{
		Kind: deep.OpRemove,
		Path: "/tasks/" + deep.EscapePathKey(taskID),
	}}}
}

// Close closes the incident — under a Guard that it is resolved first. A
// not-yet-resolved incident rejects the whole patch with a conflict rather
// than skipping: closing is not something to half-do.
func Close() deep.Patch[model.Incident] {
	return deep.NewPatch[model.Incident]().
		Guard(deep.Eq(statusPath(), model.StatusResolved)).
		With(deep.Set(statusPath(), model.StatusClosed)).
		Build()
}

// Handoff transfers command, strictly: the patch carries who the sender
// believes is in command, and if someone else took over in the meantime the
// strict Old check refuses instead of silently overriding them.
func Handoff(from, to string) deep.Patch[model.Incident] {
	return deep.Patch[model.Incident]{
		Strict: true,
		Operations: []deep.Operation{{
			Kind: deep.OpReplace,
			Path: "/commander",
			Old:  from,
			New:  to,
		}},
	}
}

File: examples/incident/web/patches.js
// The patch vocabulary, in the browser. Deliberately free of imports, so it
// is equally usable from a page, from a test, and from anything else.
//
// This is the same set of operations cmd/incident builds in Go — the same
// paths, the same conditions — because it is the same wire format. A patch
// built here is indistinguishable from one the CLI sends, which is the point
// of the exercise: the server has no idea which one it is talking to.
/** Escapes a path token, as RFC 6901 requires. */
function key(id) {
  return String(id).replaceAll('~', '~0').replaceAll('/', '~1');
}

/**
 * Claim a task — but only if nobody holds it.
 *
 * Two responders clicking at once resolve on the server: one patch applies,
 * the other *skips*, and skipping is not an error. No locking, no retry.
 */
export function claimTask(taskID, who) {
  const owner = `/tasks/${key(taskID)}/owner`;
  return {
    ops: [{ k: 'replace', p: owner, n: who, if: { p: owner, o: '==', v: '' } }],
  };
}

/** Mark a task done. */
export function completeTask(taskID) {
  return { ops: [{ k: 'replace', p: `/tasks/${key(taskID)}/done`, n: true }] };
}

/**
 * Raise severity — but never lower it. Applied twice, or after somebody
 * escalated further, this skips rather than undoing their work.
 */
export function escalate(severity) {
  return {
    ops: [{ k: 'replace', p: '/severity', n: severity, if: { p: '/severity', o: '>', v: severity } }],
  };
}

/** Move the incident's status. The server judges whether the result is legal. */
export function setStatus(status) {
  return { ops: [{ k: 'replace', p: '/status', n: status }] };
}

/**
 * Close the incident, under a guard: unless it is resolved, the whole patch
 * is refused rather than half-applied.
 */
export function close() {
  return {
    cond: { p: '/status', o: '==', v: 'resolved' },
    ops: [{ k: 'replace', p: '/status', n: 'closed' }],
  };
}

/**
 * The key schema for this model: Go marks the identity of a task with a
 * `deep:"key"` struct tag, which is invisible in JSON, so a JavaScript peer
 * has to be told. With it, `diff` addresses tasks as /tasks/<id> exactly as
 * the Go server does.
 */
export const keys = { '/tasks': 'id' };
